File: learning-factory/warehouse/artifacts/projects/build-your-own-x/programming-language/from-source-code-to-machine-code-build-your-own-compiler-from-scratch-2877aeded6/job_byox_build_v1_7a07ec294de1cc10739ceacd586b1ffd/attempt-001/starter/minnow/vm.py
"""Validated Minnow stack-machine execution."""

from .bytecode import validate


def run_bytecode(program, stdout, *, step_limit=1_000_000):
    if not isinstance(program, (bytes, bytearray)):
        raise TypeError("program must be bytes or bytearray")
    if isinstance(step_limit, bool) or not isinstance(step_limit, int):
        raise TypeError("step_limit must be an integer")
    if step_limit <= 0:
        raise ValueError("step_limit must be positive")
    verified = validate(bytes(program))
    # TODO: dispatch verified instructions with explicit signed-64 arithmetic.
    del verified, stdout
    raise NotImplementedError("implement virtual machine")

File: learning-factory/warehouse/artifacts/projects/build-your-own-x/programming-language/a-compiler-from-scratch-84427e53fe/job_byox_build_v1_f3e0d99cf97b485202190e3be4a0e8f6/attempt-001/starter/lib/pebble/vm.rb
# frozen_string_literal: true

module Pebble
  class VM
    def initialize(program, output: $stdout, max_steps: 100_000)
      @program = program
      @output = output
      @max_steps = max_steps
    end

    def run
      raise NotImplementedError, "implement Pebble::VM#run"
    end
  end
end
